File: examples/fact/fact.ml
let rec factorial n =
  match n with
  | 0 -> 1
  | 1 -> 1
  | _ -> n * factorial (n - 1)
;;

print_string "Enter a number: ";
print_newline ();
flush stdout;

try
  let number = read_int () in
  if number < 0 then
    raise (Failure "Invalid input. Please enter a non-negative integer.")
  else
    let result = factorial number in
    print_int result;
    print_newline ()
with
| Failure msg -> print_endline msg
| _ -> print_endline "Invalid input. Please enter a valid non-negative integer."

File: examples/fact/fact.py
def factorial(n: int) -> int:
    result = 1
    for n in range(1, n + 1):
        result *= n
    return result


if __name__ == "__main__":
    raw = input("enter a number:\n")

    if not raw.isdigit():
        raise Exception("Invalid input. Please enter a valid non-negative integer.")

    number = int(raw)
    if number < 0:
        raise Exception("Invalid input. Please enter a non-negative integer.")

    print(factorial(number))
